proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+  using the selected task until user changes the current task. For example:
+
+  ``` python
+	  qa = mtc.task("question-generation")
+      qa(documents=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ['What is the capital of Germany?']
+  ```
+
+- The addition of new tasks is supported by the `add_task` method. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.add_task(TaskTemplate(name="sentiment-analysis",
+                          prompt_text="Please give a sentiment for this context. Answer with positive, negative or neutral. Context: $documents; Answer:",
+                          prompt_params=["documents"]))
+      mtc.supported_task_names()
+
+      ----------------------------------------------------------------------------
+      ['question-answering',
+       'question-generation',
+       'conditioned-question-generation',
+       'summarization',
+       'question-answering-check',
+       'sentiment-analysis']
+  ```
+
+- Task parameters verification
+
+  All task parameters will be verified to match the task template and the corresponding object parameters will be
+  checked for type and value. For example:
+
+  ``` python
+      from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.prompt_for_task("question-generation", some_unknown_param=["Berlin is the capital of Germany."])
+
+      ----------------------------------------------------------------------------
+      ValueError                                Traceback (most recent call last)
+      <ipython-input-16-369cca52e960> in <module>
+            1 # tasks parameters are checked
+      ----> 2 sa(some_param=[Document("Berlin is the capital of Germany.")])
+
+      2 frames
+      /usr/local/lib/python3.8/dist-packages/haystack/nodes/llm/multi_task.py in __call__(self, *args, **kwargs)
+           34         if set(template_dict.keys()) != set(self.prompt_params):
+           35             available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+      ---> 36             raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+           37
+           38         template_dict["prompt_template"] = self.prompt_text
+
+      ValueError: Expected prompt params ['documents'] but got ['some_unknown_param']
+  ```
+
+- MultiTaskPipelines
+
+  Even though we can use MultiTaskComponent directly its real power lies in using pipelines and Haystack. For example, we
+  can retrieve documents from the document store using the query and then inject the retrieved documents into documents
+  as a parameter to the selected MultiTaskComponent task. For example:
+
+  ``` python
+      from haystack.pipelines import MultiTaskPipeline
```

Comment:
I also would like to say that we would ideally set up `MultiTaskComponent` to work natively in the existing `Pipeline` object and not introduce `MultiTaskPipeline`. It looks like from the code example below that we can get `MultiTaskComponent` to work in the standard `Pipeline` in Haystack. So I'd like to ask what are the pros for introducing `MultiTaskPipeline`?

Reply:
Good point, others voiced similar concerns